ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/BlockFetch/Client.hs

```diff
@@ -0,0 +1,424 @@
+        securityParam  = SecurityParam 1000
+        topLevelConfig = singleNodeTestConfigWithK securityParam
+        epochSize      = eraEpochSize $ topLevelConfigLedger topLevelConfig
+
+        mockedHasFS = SomeHasFS . simHasFS <$> newTVarIO MockFS.empty
+
+        cdbTracer = Tracer \case
+            ChainDBImpl.TraceAddBlockEvent ev ->
+              traceWith tracer $ "ChainDB: " <> show ev
+            _ -> pure ()
+
+        -- TODO generate more interesting behavior here
+        checkInFuture vf = pure (validatedFragment vf, [])
+
+    mkTestBlockFetchConsensusInterface ::
+         STM m (Map PeerId (AnchoredFragment (Header TestBlock)))
+      -> BlockFetchClientInterface.ChainDbView m TestBlock
+      -> BlockFetchConsensusInterface PeerId (Header TestBlock) TestBlock m
+    mkTestBlockFetchConsensusInterface getCandidates chainDbView =
+        BlockFetchClientInterface.mkBlockFetchConsensusInterface
+          (TestBlockConfig numCoreNodes)
+          chainDbView
+          getCandidates
+          (\_ -> 1000)
+          slotForgeTime
+          blockchainTime
+      where
+        -- Bogus implementation; this is fine as this is only used for
+        -- enriching tracing information ATM.
+        slotForgeTime :: BlockFetchClientInterface.SlotForgeTimeOracle m blk
+        slotForgeTime _ = pure UTCTime {
+            utctDay     = fromGregorian 1970 1 1
+          , utctDayTime = 0
+          }
+
+        -- This is only used to detect the 'FetchMode' of the
+        -- BlockFetch clients, which is not of concern for this test.
+        blockchainTime = BlockchainTime {
+            getCurrentSlot = pure $ CurrentSlot 0
+          }
+
+mockBlockFetchServer ::
+     forall m blk.
+     (Monad m, HasHeader blk)
+  => m (AnchoredFragment blk)
+  -> BlockFetchServer blk (Point blk) m ()
+mockBlockFetchServer getCurrentChain = idle
+  where
+    idle :: BlockFetchServer blk (Point blk) m ()
+    idle = flip BlockFetchServer () \(ChainRange from to) -> do
+        curChain <- getCurrentChain
+        pure case AF.sliceRange curChain from to of
+          Nothing    -> SendMsgNoBlocks (pure idle)
+          Just slice -> SendMsgStartBatch $ sendBlocks (AF.toOldestFirst slice)
+
+    sendBlocks :: [blk] -> m (BlockFetchSendBlocks blk (Point blk) m ())
+    sendBlocks = pure . \case
+      []         -> SendMsgBatchDone (pure idle)
+      blk : blks -> SendMsgBlock blk (sendBlocks blks)
+
+ntnVersion :: NodeToNodeVersion
+ntnVersion = maxBound
+
+{-------------------------------------------------------------------------------
+  BlockFetchClientSetup
+-------------------------------------------------------------------------------}
+
+data BlockFetchClientSetup = BlockFetchClientSetup {
+    -- | A 'Schedule' of 'ChainUpdate's for every peer. This emulates
+    -- the candidate fragments provided by the ChainSync client.
+    peerUpdates         :: Map PeerId (Schedule ChainUpdate)
+  }
+
+instance Show BlockFetchClientSetup where
```

Comment:
Good points, but usually, I find it nice to have a more readable output in case of failure. The derived `Show` instance can quickly be regained if necessary by just commenting out the custom one and adding a derive clause, so that possibility is not lost.

Just for comparison, the ChainSync client test also uses a "pretty" `Show` instance (though we could change that).